lib/srv/db/dynamodb/engine.go

```diff
@@ -0,0 +1,421 @@
+			DatabaseService:  e.sessionCtx.Database.GetName(),
+			DatabaseProtocol: e.sessionCtx.Database.GetProtocol(),
+			DatabaseURI:      uri,
+			DatabaseName:     e.sessionCtx.DatabaseName,
+			DatabaseUser:     e.sessionCtx.DatabaseUser,
+		},
+		StatusCode: uint32(statusCode),
+		Path:       req.URL.Path,
+		RawQuery:   req.URL.RawQuery,
+		Method:     req.Method,
+		Target:     target,
+		Body:       body,
+	}
+	e.Audit.EmitEvent(e.Context, event)
+}
+
+// checkAccess does authorization check for DynamoDB connection about
+// to be established.
+func (e *Engine) checkAccess(ctx context.Context, sessionCtx *common.Session) error {
+	ap, err := e.Auth.GetAuthPreference(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	mfaParams := sessionCtx.MFAParams(ap.GetRequireMFAType())
+	dbRoleMatchers := role.DatabaseRoleMatchers(
+		sessionCtx.Database.GetProtocol(),
+		sessionCtx.DatabaseUser,
+		sessionCtx.DatabaseName,
+	)
+	err = sessionCtx.Checker.CheckAccess(
+		sessionCtx.Database,
+		mfaParams,
+		dbRoleMatchers...,
+	)
+	return trace.Wrap(err)
+}
+
+// getRoundTripper makes an HTTP client with TLS config based on the given URI host.
+func (e *Engine) getRoundTripper(ctx context.Context, uri string) (http.RoundTripper, error) {
+	if rt, ok := e.RoundTrippers[uri]; ok {
+		return rt, nil
+	}
+	tlsConfig, err := e.Auth.GetTLSConfig(ctx, e.sessionCtx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	hostname, err := getURIHostname(uri)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	// We need to set the ServerName here because the AWS endpoint service prefix is not known in advance,
+	// and the TLS config we got does not set it.
+	tlsConfig.ServerName = hostname
+
+	out, err := defaults.Transport()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	out.TLSClientConfig = tlsConfig
+	e.RoundTrippers[uri] = out
+	return out, nil
+}
+
+// getService extracts the service ID from the request header X-Amz-Target.
+func (e *Engine) getService(req *http.Request) (string, error) {
+	// try to get the service from x-amz-target header.
+	target := req.Header.Get(libaws.AmzTargetHeader)
+	if target == "" {
+		return "", trace.BadParameter("missing %q header in http request", libaws.AmzTargetHeader)
+	}
+	service, err := parseDynamoDBServiceFromTarget(target)
+	return service, trace.Wrap(err)
+}
+
+// getTargetURI returns the target URI constructed from a given service.
+func (e *Engine) getTargetURI(service string) (string, error) {
+	uri := e.sessionCtx.Database.GetURI()
+	if !apiaws.IsAWSEndpoint(uri) {
+		// non-aws endpoints are used without modification. This is an unusual configuration but supports unforeseen use-cases.
+		return uri, nil
+	}
+	// when the database is created we ensure any AWS endpoint is configured to be a suffix missing only the service prefix.
+	prefix, err := endpointPrefixForService(service)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	return prefix + uri, nil
```

Comment:
nit: personally i would use [endpoints.DefaultResolver().EndpointFor](https://docs.aws.amazon.com/sdk-for-go/api/aws/endpoints/#Partition.EndpointFor) to create the url using the `ServiceID` and `Region`

Reply:
refactored endpoint resolution around this concept. It simplified things greatly!